Add import CI tests & various db fixes

jagerman opened the pull request:
This adds a CI job and scripts that take a v0.1.10 test database, import it, and test that we get the expected result.  It currently includes three tests: 0.1.10 -> curr (for sqlite + postgresql), and 0.1.10 -> 0.2.0 -> curr (for sqlite).

This found all sorts of upgrade issues, fixed herein.

- v0.1.x import has been non-functional for a long time (since just after the 0.2.0 tag)
- make v0.1.x import work on postgresql
- inbox was only in the migration code but missing from the master schemas
- the import hacks table had non-cascading foreign keys, which prevented deletion; this PR fixes them and adds a migration fix for existing setups (fixes #62)
- message_metadata upgrades broke seqno column renames under sqlite, and ended up with the wrong name under postgres
- sqlite inbox had incorrect 1d default expiry instead of 15d
- `citext` postgresql extension was getting put in the sogs_test schema by the postgresql test suite, which made it unusable in the public schema if you used the same db for something other that pytest.  Fixed it to always put it in the public schema.

~~Still TODO:~~
- [x] create from-scratch v0.2.0 snapshot (i.e. initial non-sogs install) for sqlite3
- [x] create from-initial-pg snapshot upgrade for postgresql

jagerman committed:
Add inbox creation to master schema files

jagerman committed:
Fix non-cascading foreign key in hack tables

The 0.1.x import was creating the two hacks table with a non-cascading
foreign key, which prevented the room from being deleted because of the
foreign key.

This makes them cascading in the import, and fixes installations with
them to be cascading.

jagerman committed:
Fix migration errors

The message_metadata table was causing upgrade errors because sqlite
rename only applies to immediate views, but not views-of-views (and
message_metadata is a view of message_details which is a view of
messages), and so the rename would fail.

This fixes it by dropping the views and recreating them later in the
import.

Also do this on postgresql, because the same scenario in postgresql
works, but doesn't rename the output column in the view which would also
cause problems.

jagerman committed:
Fix SQLite inbox expiry to match postgresql (15d instead of 1d)

jagerman committed:
Fix 0.1.x import

- db.sqlite_connect doesn't exist anymore.

- db.query usage here had bitrotted

- Apply seqno/message_sequence column renames

- Fix a broken insert_and_get_pk

jagerman committed:
Make sure citext is created in the public schema

Otherwise, if the postgresql test suite was run, it might be in
sogs_tests and not accessible (e.g. when sometimes running local
postgresql tests in the same non-ephemeral database).

jagerman committed:
Don't create a second transaction in v_0_1_x import

Each migration *already* runs inside a transaction, so we don't need to
take out another one like this.

jagerman committed:
v0.1.x postgresql compatibility fixes

- quote "user"
- use BIGINTs instead of INTEGERs (sqlite doesn't care, pg does)
- use a query rather than an ON CONFLICT to insert the user, because
  under postgresql the latter still increments the primary key value
  (which probablyu doesn't hurt, but leaves lots of unnecessary id
  gaps).

jagerman committed:
Add import test scripts and CI pipeline

This adds a CI pipeline with scripts that imports from a small v0.1.10
test database I created and ensures that we can do various import paths:

- 0.1.10 -> current (sqlite3)
- 0.1.10 -> 0.2.0 -> current (sqlite3)
- 0.1.10 -> current (postgresql)

As we make additional releases these paths will grow.

jagerman committed:
Add whisper_mods to message_metadata view

jagerman committed:
Accommodate v0.2.0 import over-incrementing message_sequence

v0.2.0 had a trigger that added an extra (unnecessary) increment to
message_sequence (which was then called `updates`).  It didn't hurt
anything, but if we import via such a version we breaks the import tests
because the message_sequence value will be one too high, so hack it back
down by one to make it work.

majestrate commented on sogs/migrations/v_0_1_x.py:
```diff
@@ -27,6 +27,14 @@ def migrate(conn):
+
+    conn = sqlite3.connect(f"file:{path}?mode=ro", uri=True, isolation_level=None)
+    conn.row_factory = sqlite3.Row
+    return conn
```
since this is used in a `with` statement does returning here work instead of using a `yield`?

jagerman committed:
Reformat output

- Don't report recent_upload for non-expiring uploads (i.e. room images)

- Reformat because python3-tabulate module is stupid when a column has
  rows + numbers and tries to guess by trying to parsing all strings as
  numbers, WTF.

jagerman committed:
Add 0.2.0 upgrade test

jagerman committed:
Add pg upgrade tests

This starts from the dev commit just before this PR because there were
bugs setting up a new sogs install that were only just fixed on dev that
predate working pgsql support.

majestrate reviewed (commented):
nitpick: the urls from oxen.rocks should probably be in an /assets/ sub dir

jagerman committed:
Put sogs test bundles in /sogs-assets/ oxen.rocks directory

jagerman commented:
> nitpick: the urls from oxen.rocks should probably be in an /assets/ sub dir

Done.

jagerman committed:
Add missing CI dep

jagerman committed:
Replace tabulate with prettytable

tabulate is crap.  Replace it with prettytable which is much more sane.
